### DISCUSSION TITLE
Ignore nodes in semantic actions

### DISCUSSION THREAD

**Discussion by FredericEspiau** [ohmjs/ohm#333] [Help] [answered]
Hey everyone,

I love `Ohm`, thank you for all your hard work !

I do have a question: is there a way to ignore nodes in a semantic action ?

For example in the `lisp` example: https://github.com/harc/ohm/blob/master/examples/simple-lisp/src/simple-lisp.mjs#L294

It would be great to be able to write

```ts
argList(args) {
  return args.asIteration().children.map(c => c.toAST());
},
```

instead of

```ts
argList(_1, args, _2) {
  return args.asIteration().children.map(c => c.toAST());
},
```

Thanks for your time !

**Comment by pdubroy** [comment, marked as answer]
Hi @FredericEspiau,

> I love `Ohm`, thank you for all your hard work !

Nice to hear that. You're welcome 😊

> I do have a question: is there a way to ignore nodes in a semantic action ?

No, there's no way to do that currently.

**Reply by FredericEspiau** [reply]
Thanks for the quick answer !


I don't know how hard it would be to implement this. If it's easy enough, would you mind if I do a PR to add this feature ? 

I can see that `!` isn't used yet as an operator

https://github.com/harc/ohm/blob/master/packages/ohm-js/src/ohm-grammar.ohm#L119

Which would mean that the lisp example would look like this for example:

```
Sexp = !"(" Stmt !")"
```

What do you think ?

**Reply by pdubroy** [reply]
The main reason that we don't currently support something like this is that we try to maintain a clear separation between grammar and semantics. A `!` operator like this seems like it could result in grammars which are tailored to a specific semantics, which is something that we'd like to avoid. Plus, I'd prefer not to clutter grammars with symbols that aren't related to the language definiton.

What do you see as the main advantages of doing something like this?

**Reply by FredericEspiau** [reply]
This is a really minor issue as semantic action declarations can get messy when you've got a lot of boilerplate, for example the `import` statement in Typescript

```ts
import module3, { module1, module2 } from "module4";
```

I would have to deal in my parameters with `import`, `{`, `}`, `from` and `"`, even though I absolutely don't need them, they are only there for matching purpose. And splitting into multiple actions wouldn't help much as the parameters would still be there, only in different places

But I would understand if you wanted to keep it this way :)
